File: my/src/types/library.ts
// 素材库（网盘 + 容量套餐）前端类型。
// 契约源：backend/modules/library/internal/{overview.go, files.go, folders.go, tags.go,
//   pricingconfig_model.go, service.go}、billing 的 bizorder_model.go（params）。
// 金额一律 cents（分）；时间一律 RFC3339 字符串；bps：10000=原价无折扣，9500=9.5折。

// ---- 文件类型枚举（后端 classifyFileType） ----
export type FileType = 'image' | 'video' | 'audio' | 'document' | 'app' | 'other'
export type FileStatus = 'uploading' | 'active' | 'deleted'

// ---- 套餐动作（后端 ActionXxx / biz_type lib_*） ----
export type PackageAction = 'new' | 'renew' | 'upgrade' | 'downgrade'
export type LibraryBizType = 'lib_new' | 'lib_renew' | 'lib_upgrade' | 'lib_downgrade'

// ---- 定价配置档位（pricingconfig_model.go） ----
export interface TierCfg {
  code: string // t50 / t100 / t500
  name: string
  capacity_gb: number
  monthly_price_cents: number
  tier_discount_bps: number // 10000=原价
  enabled: boolean
  sort: number
}
export interface CustomTierCfg {
  enabled: boolean
  min_capacity_gb: number
  price_per_gb_month_cents: number
  tier_discount_bps: number
}
export interface DurationOptCfg {
  days: number
  discount_bps: number
}

// ---- GET /library/overview ----
export interface OverviewUsage {
  used_bytes: number
  capacity_bytes: number
  locked: boolean // used_bytes > capacity_bytes → 超额锁定
}
export interface OverviewSubscription {
  tier_code: string // free 或档位码
  capacity_bytes: number
  monthly_price_cents: number
  expire_at: string | null // 免费=null 永不过期
  is_free: boolean
}
export interface LibraryOverview {
  usage: OverviewUsage
  subscription: OverviewSubscription
  free_quota_bytes: number
  tiers: TierCfg[]
  custom_tier: CustomTierCfg
  duration_options: DurationOptCfg[]
  notice: string
  billing_note: string
}

// ---- 文件 ----
export interface LibraryFile {
  id: number
  folder_id: number // 0=根
  name: string
  ext: string
  mime_type: string
  file_type: FileType
  size_bytes: number
  status: FileStatus
  md5: string
  created_at: string
  updated_at: string
  // ListFiles 随单返回标签 ID 列表（FileDTO）。
  tag_ids?: number[]
}

// ---- 文件夹 ----
export interface LibraryFolder {
  id: number
  parent_id: number // 0=根
  name: string
  created_at: string
  updated_at: string
}

// ---- 标签 ----
export interface LibraryTag {
  id: number
  name: string
  color: string
  created_at: string
}

// ---- 上传两段式 ----
export interface PresignUploadReq {
  name: string
  size_bytes: number
  mime: string
  folder_id: number
  md5: string // 全量 md5（全局去重键之一）
  slice_md5: string // 前 256KB 的 md5（秒传持有证明）
}
export interface PresignUploadResult {
  file_id: number
  // 秒传命中（instant=true）时后端不返回直传地址，前端跳过 PUT/confirm。
  upload_url?: string
  s3_key?: string
  expires_in?: number // 秒
  instant: boolean // true=秒传命中，文件已 active
}
export interface ConfirmUploadReq {
  file_id: number
  tag_ids?: number[]
}

// ---- 改名 / 移动 ----
export interface UpdateFileReq {
  name?: string
  folder_id?: number
}

// ---- 文件列表查询 ----
export interface FileListQuery {
  folder_id?: number
  file_type?: FileType | ''
  tag_id?: number
  keyword?: string
  page?: number
  size?: number
}

// ---- 套餐报价（POST /library/package/quote） ----
// 请求载荷（service.go PackageParams）：预设档带 tier_code；自定义档带 capacity_gb；
// upgrade/downgrade 不带 days。action 可留空，由服务端按月价自动判定。
export interface PackageParams {
  action?: PackageAction | ''
  tier_code?: string
  capacity_gb?: number
  days?: number
}
// 响应（service.go QuoteResult）。
export interface PackageQuoteResult {
  action: PackageAction
  tier_code: string
  tier_name: string
  capacity_bytes: number
  days: number
  monthly_price_cents: number
  tier_discount_bps: number
  duration_discount_bps: number
  remaining_days: number
  total_cents: number
  new_expire_at: string | null
}

// ---- 下单 / 支付（直接打 billing，携带 params） ----
export type OrderStatus = 'unpaid' | 'paid' | 'expired'
export interface PackageOrderCreateReq {
  biz_type: LibraryBizType
  pay_method: string
  params: PackageParams
}
export interface LibraryOrder {
  id: number
  biz_type: string
  status: OrderStatus
  total_cents: number
  fee_cents: number
  pay_method: string
  created_at: string
  paid_at: string | null
}
export interface PayInfo { status: OrderStatus, [k: string]: unknown }
export interface OrderCreateResult { order: LibraryOrder, pay: PayInfo }

File: backend/modules/library/library.go
// Package library 是素材库业务模块的公开入口。
//
// 模块实现位于 modules/library/internal，受 Go internal 机制保护；公开包暴露"取文件"对外接口
// （OpenFile / ListUsableFiles，§7），供后续"推送到手机""脚本输入参数"等消费场景使用——
// 属主校验与超额锁定逻辑只在 internal 一处，跨模块只走本门面。
package library

import (
	"io"
	"time"

	libraryinternal "manager-backend/modules/library/internal"

	"gorm.io/gorm"
)

// FileRef 一个可用文件的引用（含短时有效的 presigned GET URL）。
type FileRef struct {
	FileID    uint
	Name      string
	MimeType  string
	SizeBytes int64
	URL       string
}

// FileContentRef 一个可用文件的内容流（调用方负责 Close Reader）。
type FileContentRef struct {
	FileID    uint
	Name      string
	MimeType  string
	SizeBytes int64
	Reader    io.ReadCloser
}

// UsableFilter 消费侧选择器过滤（按类型/标签/关键字）。
type UsableFilter struct {
	FileType string // image|video|audio|document|other；空=不限
	TagID    uint   // >0 时按标签过滤
	Keyword  string
	Page     int
	Size     int
}

// OpenFile 取一个可用文件：内部统一做属主校验 + 超额锁定校验 → presigned GET。
func OpenFile(userID int, fileID uint) (FileRef, error) {
	r, err := libraryinternal.Service.OpenFile(userID, fileID)
	if err != nil {
		return FileRef{}, err
	}
	return FileRef(r), nil
}

// OpenFileWithTTL 取一个可用文件，但 presigned GET URL 用调用方指定的 ttl（而非默认 GET TTL）。
// 校验与 OpenFile 完全一致（属主校验 + 超额锁定校验 + 仅 active）。
// 用于应用按 URL 安装等需更长有效期、够中台下载完成的场景（见 §7.4）。
func OpenFileWithTTL(userID int, fileID uint, ttl time.Duration) (FileRef, error) {
	r, err := libraryinternal.Service.OpenFileWithTTL(userID, fileID, ttl)
	if err != nil {
		return FileRef{}, err
	}
	return FileRef(r), nil
}

// OpenFileContent 取一个可用文件的字节流：内部统一做属主校验 + 超额锁定校验 + 仅 active → S3 GetObject。
// 私有桶字节直读直传（不经浏览器、不二次 HTTP presigned URL）。调用方负责 Close 返回的 Reader。
func OpenFileContent(userID int, fileID uint) (FileContentRef, error) {
	r, err := libraryinternal.Service.OpenFileContent(userID, fileID)
	if err != nil {
		return FileContentRef{}, err
	}
	return FileContentRef(r), nil
}

// DeleteFileForUser 删除某属主的素材库文件：内部做属主校验 + 软删 + 回收 used_bytes（释放配额）。
// 供应用管理「删素材库文件即应用消失」等消费侧使用——属主与配额逻辑仍只在 internal 一处。
func DeleteFileForUser(ownerUserID int, fileID uint) error {
	return libraryinternal.Service.DeleteFile(ownerUserID, fileID)
}

// ListUsableFiles 列出可用文件（带类型/标签过滤），供消费侧选择器使用。
func ListUsableFiles(userID int, filter UsableFilter) ([]FileRef, error) {
	list, err := libraryinternal.Service.ListUsableFiles(userID, libraryinternal.UsableFilter{
		FileType: filter.FileType,
		TagID:    filter.TagID,
		Keyword:  filter.Keyword,
		Page:     filter.Page,
		Size:     filter.Size,
	})
	if err != nil {
		return nil, err
	}
	out := make([]FileRef, 0, len(list))
	for _, r := range list {
		out = append(out, FileRef(r))
	}
	return out, nil
}

// InitForTest 测试用：供其他模块装配 library（建表 + 装配服务 + 注册 billing 业务类型）。
func InitForTest(db *gorm.DB) error { return libraryinternal.InitForTest(db) }
